File: remote/src/main/java/com/kelsos/mbrc/presenters/BrowseGenrePresenterImpl.java
package com.kelsos.mbrc.presenters;

import com.google.inject.Inject;
import com.kelsos.mbrc.annotations.MetaDataType;
import com.kelsos.mbrc.annotations.Queue;
import com.kelsos.mbrc.constants.Constants;
import com.kelsos.mbrc.domain.Genre;
import com.kelsos.mbrc.interactors.QueueInteractor;
import com.kelsos.mbrc.interactors.library.LibraryGenreInteractor;
import com.kelsos.mbrc.interactors.playlists.PlaylistAddInteractor;
import com.kelsos.mbrc.ui.views.BrowseGenreView;
import roboguice.util.Ln;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

public class BrowseGenrePresenterImpl implements BrowseGenrePresenter {
  private BrowseGenreView view;
  @Inject private LibraryGenreInteractor interactor;
  @Inject private QueueInteractor queue;
  @Inject private PlaylistAddInteractor playlistAddInteractor;

  @Override public void bind(BrowseGenreView view) {
    this.view = view;
  }

  @Override public void load() {
    interactor.execute(0, Constants.PAGE_SIZE)
        .subscribeOn(Schedulers.io())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(genres -> {
          view.clear();
          view.update(genres);
        }, Ln::v);
  }

  @Override public void queue(Genre genre, @Queue.Action String action) {
    queue.execute(MetaDataType.GENRE, action, (int) genre.getId())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(success -> {
          if (success) {
            view.showEnqueueSuccess();
          } else {
            view.showEnqueueFailure();
          }
        }, throwable -> {
          view.showEnqueueFailure();
        });
  }

  @Override public void load(int page) {
    interactor.execute(page * Constants.PAGE_SIZE, Constants.PAGE_SIZE)
        .subscribeOn(Schedulers.io())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(genres -> {
          view.update(genres);
        }, Ln::v);
  }

  @Override public void createPlaylist(long selectionId, String name) {

  }

  @Override public void playlistAdd(long selectionId, long playlistId) {

  }
}

File: app/src/main/kotlin/com/kelsos/mbrc/presenters/BrowseGenrePresenterImpl.kt
package com.kelsos.mbrc.presenters

import com.kelsos.mbrc.annotations.MetaDataType
import com.kelsos.mbrc.annotations.Queue
import com.kelsos.mbrc.constants.Constants
import com.kelsos.mbrc.domain.Genre
import com.kelsos.mbrc.extensions.task
import com.kelsos.mbrc.interactors.QueueInteractor
import com.kelsos.mbrc.interactors.library.LibraryGenreInteractor
import com.kelsos.mbrc.interactors.playlists.PlaylistAddInteractor
import com.kelsos.mbrc.ui.views.BrowseGenreView
import timber.log.Timber
import javax.inject.Inject

class BrowseGenrePresenterImpl
@Inject constructor(private val interactor: LibraryGenreInteractor,
                    private val queue: QueueInteractor,
                    private val playlistAddInteractor: PlaylistAddInteractor) : BrowseGenrePresenter {
  private var view: BrowseGenreView? = null

  override fun bind(view: BrowseGenreView) {
    this.view = view
  }

  override fun load() {
    interactor.execute().task().subscribe({
      view?.clear()
      view?.update(it)
    }, { Timber.v(it, "Failed to load") })
  }

  override fun queue(genre: Genre, @Queue.Action action: String) {
    queue.execute(MetaDataType.GENRE,
        action,
        genre.id).task().subscribe({
      if (it!!) {
        view?.showEnqueueSuccess()
      } else {
        view?.showEnqueueFailure()
      }
    }) { view!!.showEnqueueFailure() }
  }

  override fun load(page: Int) {
    interactor.execute(page * Constants.PAGE_SIZE)
        .task()
        .subscribe({ view!!.update(it) }, { Timber.v(it) })
  }

  override fun createPlaylist(selectionId: Long, name: String) {

  }

  override fun playlistAdd(selectionId: Long, playlistId: Long) {

  }
}
